src/priority-queue.ts

```diff
@@ -0,0 +1,262 @@
+		const parentIndex = Math.floor((nodeIndex - 1) / 2);
+		if (parentIndex >= this.size()) {
+			return null;
+		}
+
+		return parentIndex;
+	}
+
+	public getChildNodesIndexes(nodeIndex: number | null): [number | null, number | null] {
+		if (typeof nodeIndex !== 'number') {
+			return [null, null];
+		}
+		if (nodeIndex < 0) {
+			return [null, null];
+		}
+		if (nodeIndex >= this.size()) {
+			return [null, null];
+		}
+		if (nodeIndex % 1 !== 0) {
+			return [null, null];
+		}
+
+		const childOneIndex = nodeIndex * 2 + 1;
+		const childTwoIndex = nodeIndex * 2 + 2;
+		if (childOneIndex >= this.size()) {
+			return [null, null];
+		}
+		if (childTwoIndex >= this.size()) {
+			return [childOneIndex, null];
+		}
+
+		return [childOneIndex, childTwoIndex];
+	}
+
+	public getNextIndex(startFromTop: boolean, nodeIndex: number | null): number | null {
+		if (!startFromTop) {
+			return this.getParentNodeIndex(nodeIndex);
+		}
+
+		const [leftIndex, rightIndex] = this.getChildNodesIndexes(nodeIndex);
+		if (leftIndex === null || rightIndex === null) {
+			return leftIndex;
+		}
+		if (this.elements[leftIndex] === null) {
+			return rightIndex;
+		}
+		if (this.elements[rightIndex] === null) {
+			return leftIndex;
+		}
+
+		if (this.comparator(leftIndex, rightIndex)) {
+			return leftIndex;
+		} else {
+			return rightIndex;
+		}
+	}
+
+	public isHeapUnbalanced(nodeIndex: number | null, nextIndex: number | null): Boolean {
+		if (typeof nodeIndex !== 'number') {
+			return false;
+		}
+		if (typeof nextIndex !== 'number') {
+			return false;
+		}
+
+		const nodeValue = this.elements[nodeIndex];
+		const nextValue = this.elements[nextIndex];
+		const startFromTop = nodeIndex < nextIndex;
+
+		if (nodeValue === null) {
+			return false;
+		}
+		if (nextValue === null) {
+			return false;
+		}
+
+		// i think this is superfluous, but not deleting until testing is complete
+		// if (!startFromTop && nodeIndex <= 0) {
+		// 	console.log(nodeIndex,nextIndex);
+		// 	return false;
+		// }
+
+		if (startFromTop) {
+			return this.comparator(nextIndex, nodeIndex);
+		} else {
+			return this.comparator(nodeIndex, nextIndex);
+		}
+	}
+
+	public fixHeap(nodeIndex: number | null): void {
+		if (this.size() <= 1) {
+			return;
+		}
+		if (typeof nodeIndex !== 'number') {
+			return;
+		}
+		if (nodeIndex < 0) {
+			return;
+		}
+		if (nodeIndex >= this.size()) {
+			return;
+		}
+		if (nodeIndex % 1 !== 0) {
+			return;
+		}
+
+		const startFromTop = nodeIndex === 0;
+		let nextIndex = this.getNextIndex(startFromTop, nodeIndex);
+
+		while (this.isHeapUnbalanced(nodeIndex, nextIndex)) {
+			this.swapNodes(nodeIndex, nextIndex);
+			nodeIndex = nextIndex;
+			nextIndex = this.getNextIndex(startFromTop, nodeIndex);
+		}
+	}
+
+	public push(element: T): ArmorPriorityQueue<T> {
+		this.elements.push(element);
+		this.fixHeap(this.size() - 1);
+		return this;
+	}
+
+	public pop(): T | null {
+		if (this.size() === 0) {
+			return null;
+		}
+		if (this.size() === 1) {
```

Comment:
My initial instinct is pop will be faster than creating and assigning a new array. Unfortunately whether it matters at all depends how frequently a single element array is popped. If you have data on this being a faster optimization we should keep it, otherwise we can simplify this short block down to:
```typescript
return this.elements.pop();
```